pageserver/src/restore_local_repo.rs

```diff
@@ -426,14 +425,38 @@ pub fn save_decoded_record(
             relnode: blk.rnode_relnode,
             forknum: blk.forknum as u8,
         });
-
-        let rec = WALRecord {
-            will_init: blk.will_init || blk.apply_image,
-            rec: recdata.clone(),
-            main_data_offset: decoded.main_data_offset as u32,
-        };
-
-        timeline.put_wal_record(lsn, tag, blk.blkno, rec)?;
+        if blk.apply_image
+            && blk.has_image
+            && decoded.xl_rmid == pg_constants::RM_XLOG_ID
+            && (decoded.xl_info == pg_constants::XLOG_FPI
+                || decoded.xl_info == pg_constants::XLOG_FPI_FOR_HINT)
+        {
+            // Extract page image from FPI record
+            let img_len = blk.bimg_len as usize;
+            let img_offs = blk.bimg_offset as usize;
+            let mut image = BytesMut::with_capacity(pg_constants::BLCKSZ as usize);
+            image.extend_from_slice(&recdata[img_offs..img_offs + img_len]);
+
+            // Compression of WAL is not yet supported
+            assert!((blk.bimg_info & pg_constants::BKPIMAGE_IS_COMPRESSED) == 0);
```

Comment:
Would it be hard to support it? We don't enable WAL compression at the moment, but we might want to do that in the future. And even if we don't, it'd be good to be able to digest WAL generated with vanilla PostgreSQL, e.g. if you're importing an existing PostgreSQL installation.

As a quick fix, perhaps check for this in the if-clause above, and fall back to the old way if the image is compressed.

Reply:
Sorry, I have applied changes requested in PR #880 in wrong branch, so them were not included in commit.

I have added fallback for compressed images. 

Concerning support of WAL compression, it seems to be not so difficult, but requite some "hacks" with FFI (to be able to call Postgres pglz from Rust).